src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/SessionPass.php

```diff
@@ -22,16 +22,29 @@ class SessionPass implements CompilerPassInterface
 {
     public function process(ContainerBuilder $container)
     {
-        if (!$container->hasDefinition('session')) {
+        if (!$container->has('session')) {
             return;
         }
 
+        // BC layer: When user overrides the `session` service it's not an alias anymore.
+        if ($container->hasDefinition('session')) {
+            $definition = $container->getDefinition('session');
+            $definition->setDeprecated('symfony/framework-bundle', '5.2', 'The "%service_id%" service is deprecated, use "$requestStack->getSession()" instead.');
+
+            // Given `session` is not an alias to `.session.do-not-use` anymore,
+            // we make `.session.do-not-use` an alias of `session`.
+            $container->removeDefinition('.session.do-not-use');
+            $container->setAlias('.session.do-not-use', 'session');
+        } else {
+            $definition = $container->getDefinition('.session.do-not-use');
+        }
+
         $bags = [
             'session.flash_bag' => $container->hasDefinition('session.flash_bag') ? $container->getDefinition('session.flash_bag') : null,
             'session.attribute_bag' => $container->hasDefinition('session.attribute_bag') ? $container->getDefinition('session.attribute_bag') : null,
         ];
 
-        foreach ($container->getDefinition('session')->getArguments() as $v) {
```

Comment:
im not really sure what below is about, but depending on the above decision (basically `has` vs `hasDefinition` to detect a user defined session service) we could merge the definition arguments from both `session` and `.session.do-not-use`, if both are distinctive definitions

Reply:
If user defined it own `session` service, it should be compatible with the old way. Merging the `session.do-not-use` may break things. I prefer keeping the user's definition

Reply:
i meant merging the arguments once, so both definitions are taken into account here. BUT it's fine actually :) I was again confused, as the reversed alias (`.session.do-not-use => session`) already ensures 1 definition.

However, related to above it could work like

```php
if $container->has('session')
   $def = $container->findDef('session');
else
   $def = new Def(); // build at compile time
```

Reply:
`if $container->has('session')` returns always true, because search for Definition OR Alias

Reply:
well yes, because the user may override the session service in both ways. Either re-define or create some arbitrary alias.

Currently we cannot distinct between user-defined vs. core-defined `session` service due `->alias('session', '.session.do-not-use')` in our own config. If we move that to the else case above, `has('session')` will always indicate "user-defined".

Reply:
removed service `session` from config and check in compilerPass when overrode by user, and create alias according.